Matter TemperatureNumber feature from TemperatureControl cluster

lboue opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
It would be useful to add support for Matter [Temperature Control Cluster](https://github.com/project-chip/connectedhomeip/blob/master/src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml)  (0x0056) used in the following device types: 
* [Dishwasher](https://github.com/project-chip/connectedhomeip/blob/79c1cf5cb093448d1eb6ee9b5f31c68b75c241c8/data_model/1.3/device_types/Dishwasher.xml)
* [LaundryDryer](https://github.com/project-chip/connectedhomeip/blob/79c1cf5cb093448d1eb6ee9b5f31c68b75c241c8/data_model/1.3/device_types/LaundryDryer.xml)
* [LaundryWasher](https://github.com/project-chip/connectedhomeip/blob/79c1cf5cb093448d1eb6ee9b5f31c68b75c241c8/data_model/1.3/device_types/LaundryWasher.xml)
* [Refrigerator](https://github.com/project-chip/connectedhomeip/blob/79c1cf5cb093448d1eb6ee9b5f31c68b75c241c8/data_model/1.3/device_types/Refrigerator.xml)
* [CookSurface](https://github.com/project-chip/connectedhomeip/blob/79c1cf5cb093448d1eb6ee9b5f31c68b75c241c8/data_model/1.3/device_types/CookSurface.xml)

This would allow manufacturers to test their products.

This cluster provides an interface to the setpoint temperature on devices such as washers, refrigerators, and water heaters. The setpoint temperature is the temperature to which a device using this cluster would attempt to control to. This cluster does not provide access to the actual or physical temperature associated with any device using this cluster. Access to the physical temperature associated with a device using this cluster would be provided by other clusters as part of that devices
device type definition.

There is 3 features for this cluster:
* **TemperatureNumber** feature: 
       For devices that use an actual temperature value for the temperature setpoint, such as some water heaters
  * **TemperatureStep** feature:
       For devices that support discrete temperature setpoints that are larger than the temperature resolution imposed via the temperature data type
* **TemperatureLevel** feature: 
       For  devices that use vendor-specific temperature levels for the temperature setpoint, such as some washers (Dishwasher, LaundryDryer, LaundryWasher)


There is one command to change the target Temperature:
* **TemperatureNumber** feature: 
```
chip-tool temperaturecontrol set-temperature 1 1 --TargetTemperature 40
```

* **TemperatureLevel** feature: 
```
chip-tool temperaturecontrol set-temperature 1 1 --TargetTemperatureLevel 1
```
SupportedTemperatureLevels (86/5) is a `list[Str]`
![image](https://github.com/user-attachments/assets/103c262f-3dcc-424b-b1b0-8c6e38862f73)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

![image](https://github.com/user-attachments/assets/21009e14-f18a-486f-a611-e90dc37d589f)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

lboue committed:
Create temperature_control.py

lboue committed:
Update temperature_control.py

lboue committed:
Update temperature_control.py

lboue committed:
Update temperature_control.py

lboue committed:
Update temperature_control.py

lboue committed:
Update temperature_control.py

lboue committed:
Update temperature_control.py

lboue commented:
Is there a HomeAssistant entity to map this cluster with?

lboue committed:
ruff-format

lboue committed:
Update temperature_control.py

bouwew commented:
You should add/integrate this code into the `climate.py` file.

lboue committed:
Update temperature_control.py

lboue committed:
Update temperature_control.py

lboue committed:
TemperatureControl SelectedTemperatureLevel

lboue committed:
Update climate.py

lboue committed:
TemperatureControlFeature

lboue committed:
Delete homeassistant/components/matter/temperature_control.py

lboue committed:
TemperatureControl mode

lboue committed:
StrEnum

lboue committed:
Update climate.py

lboue committed:
Update climate.py

lboue committed:
Update climate.py

lboue committed:
Use feature name instead of feature code

lboue committed:
Update climate.py

lboue committed:
Update climate.py

lboue committed:
Create silabs_laundrywasher.json

lboue committed:
Update silabs_laundrywasher.json

lboue commented:
> You should add/integrate this code into the `climate.py` file.

For LaundryWasher device type, the spec only allow to set TemperatureLevel so it's not really a Thermostat. It doesn't expose the currentTemp

bouwew commented:
@lboue Sorry, just now I see the full scope of your PR! These devices are indeed not all climate-types!

Unfortunately, HA only has the `Climate`, `Fan`, `Humifier` and `Water heater` entities defined that might suit to what you want to implement. If not, you would need to request to add a new Entity that suits the Appliances that you want to cover in this PR.

Are you familiar with the HA Developer Docs? Here's a link to the Climate Entity: https://developers.home-assistant.io/docs/core/entity/climate

lboue commented:
> @lboue Unfortunately, HA only has the `Climate`, `Fan`, `Humifier` and `Water heater` entities defined that might suit to what you want to implement. If not, you would need to request to add a new Entity that suits the Appliances that you want to cover in this PR.

You'd need to add a new 'Temperature Control' entity to HomeAssistant beforehand.

> Are you familiar with the HA Developer Docs? Here's a link to the Climate Entity: https://developers.home-assistant.io/docs/core/entity/climate

Yes I know that page.

bouwew commented:
@lboue Please go here: https://github.com/home-assistant/architecture and search in Discussions if a similar topic exists.
If not, create a new topic :)

lboue commented:
> @lboue Please go here: https://github.com/home-assistant/architecture and search in Discussions if a similar topic exists. If not, create a new topic :)

Thanks

marcelveldt commented:
This needs a little bit more discussion to which existing platform(s) this can be fitted into (a new platform would be a last resort if really needed!). So far we managed to fit the new device types introduced in Matter into existing platforms, I'd like to keep that trend as long as we can.

I will have a look at the spec about this particular cluster (and the device types using it) to form a better opinion.

lboue committed:
Merge branch 'dev' into TemperatureControlled

lboue commented:
> This needs a little bit more discussion to which existing platform(s) this can be fitted into (a new platform would be a last resort if really needed!). So far we managed to fit the new device types introduced in Matter into existing platforms, I'd like to keep that trend as long as we can.
> 
> I will have a look at the spec about this particular cluster (and the device types using it) to form a better opinion.

Were you able to consult the documentation to give us your opinion about TemperatureControl cluster?

lboue committed:
Update silabs_laundrywasher.json

lboue committed:
Remove the select part

lboue committed:
Fixture

lboue committed:
TemperatureControl / TemperatureSetpoint

lboue committed:
Delete tests/components/matter/fixtures/nodes/silabs_laundrywasher.json

lboue commented:
I have split the implementation into 2 entities:

| Device Type                    | Cluster            | Feature           | Map              |
|--------------------------------|--------------------|-------------------|------------------|
| Temperature Controlled Cabinet | TemperatureControl | TemperatureNumber | Platform.CLIMATE |
| Temperature Controlled Cabinet | TemperatureControl |  TemperatureLevel |  Platform.SELECT |

For `TemperatureNumber` feature, I managed to make the [Silabs Refrigerator and Temperature Controlled Example](https://github.com/project-chip/connectedhomeip/tree/master/examples/refrigerator-app/silabs) work.

The other PR:
- #134532

lboue committed:
Tests for TemperatureControl cluster

lboue committed:
Update climate.py

lboue committed:
Add fixture to conftest

lboue committed:
Update tests

lboue committed:
Update tests

marcelveldt commented:
Put some thinking into this one and its a difficult case. Basically its a thermostat without a current temperature.
From a Matter perspective it is primary designed for things like fridges or washing machines, a target level you have set.

Painpoint is that this is not strictly a numeric setting, so imo we should not try to catch this into a single entity model.
For devices that are very close to an actual thermostat(like) behavior we could use climate platform but otherwise this is simply a select entity and perhaps some surrounding entities like button, number etc. to fulfil the more advabced features in the cluster.

Now the issue about having climate/thermostat without an actual current thermostat level (so only target level is known) - That is not something unique, other integrations (e.g. Z-Wave) are also familiar with those devices and our entity model should be able to handle that.

Basically this is what you already figured out and you split off the non-numeric temperature control into an enum entity.
So imo this is good the way this is now.

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -422,6 +423,87 @@ def _get_temperature_in_degrees(
 
 
+class MatterTemperatureControlClimate(MatterEntity, ClimateEntity):
+    """Representation of a Matter climate entity."""
```
```suggestion
    """Representation of a climate entity from the Matter TemperatureControl cluster."""
```

lboue committed:
Update homeassistant/components/matter/climate.py

Update comments

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -422,6 +423,87 @@ def _get_temperature_in_degrees(
+    _attr_temperature_unit: str = UnitOfTemperature.CELSIUS
+    _attr_hvac_mode: HVACMode = HVACMode.HEAT_COOL
+    _attr_hvac_modes: list[HVACMode] = [HVACMode.HEAT_COOL]
+    _attr_target_temperature_step: float | None
```
If this is already defined in the base model, no need to repeat here

lboue replied:
Fixed

lboue replied:
Without this, I am facing an error:
`AttributeError: 'MatterTemperatureControlClimate' object has no attribute '__attr_hvac_mode'. Did you mean: '__attr_hvac_modes'?`

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -422,6 +423,87 @@ def _get_temperature_in_degrees(
+    _attr_target_temperature_step: float | None
+    _feature_map: int | None = None
+
+    _platform_translation_key = "thermostat"
```
This is the thing I'm in doubt about. We could choose to define a translation key for "Temperature Control" instead of the default Thermostat label. What do you think ?

lboue replied:
Fixed

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -422,6 +423,87 @@ def _get_temperature_in_degrees(
+        target_temperature: float | None = kwargs.get(ATTR_TEMPERATURE)
+
+        if target_hvac_mode is not None:
+            await self.async_set_hvac_mode(target_hvac_mode)
```
Please add a small test case where the mode is also set in the set temperature

lboue replied:
`async_set_hvac_mode` is not implemented. This is not `Thermostat `cluster but `TemperatureControl` cluster.
We can't change mode.

marcelveldt replied:
Just leave it the references for the set mode then

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -422,6 +423,87 @@ def _get_temperature_in_degrees(
+    def _update_from_device(self) -> None:
+        """Update from device."""
+        self._calculate_features()
+
```
```suggestion
```

lboue committed:
Update homeassistant/components/matter/climate.py

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -422,6 +423,87 @@ def _get_temperature_in_degrees(
+                clusters.TemperatureControl.Attributes.FeatureMap
+            )
+        )
+        self._feature_map = feature_map
```
Is it needed to calculate the features every time ? Maybe just store the previous value of the featuremap into a variable so you can compare that to determine if the features changed, see some of the other platforms for examples

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -448,4 +530,19 @@ def _get_temperature_in_degrees(
+        ),
+        entity_class=MatterTemperatureControlClimate,
+        required_attributes=(
+            clusters.TemperatureControl.Attributes.FeatureMap,
```
Please list the primary attribute first, which is not the featuremap but the setpoint

lboue committed:
Merge branch 'dev' into TemperatureControlled

lboue committed:
Set  TemperatureSetpoint as primary attribute

lboue committed:
Only update featuremap if it changes

marcelveldt commented:
@lboue can you address the final feedback, rebase the PR and fix any merge conflicts ?
When you're ready, press the "ready for review" button for the (final) review. Thanks!

lboue committed:
Change platform_translation_key

lboue committed:
Cleanup

lboue committed:
Mandatory attributes

lboue committed:
Update tests

lboue committed:
Update from dev

lboue committed:
Merge branch 'dev' into TemperatureControlled

lboue committed:
Update snapshots

lboue commented:
It looks like we have another issue with snapshot file test_select.ambr regarding those entities:
- select.laundrywasher_number_of_rinses
- select.laundry_washer_spin_speed

They are modified when the snapshot is updated, whereas the modified mode does not refer to them.

lboue committed:
Update snapshot

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -408,6 +409,89 @@ def _get_temperature_in_degrees(
+        target_hvac_mode: HVACMode | None = kwargs.get(ATTR_HVAC_MODE)
+        target_temperature: float | None = kwargs.get(ATTR_TEMPERATURE)
+
+        if target_hvac_mode is not None:
+            await self.async_set_hvac_mode(target_hvac_mode)
```
```suggestion
```

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -408,6 +409,89 @@ def _get_temperature_in_degrees(
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
+        """Set new target temperature."""
+        target_hvac_mode: HVACMode | None = kwargs.get(ATTR_HVAC_MODE)
```
```suggestion
```

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -408,6 +409,89 @@ def _get_temperature_in_degrees(
+            await self.async_set_hvac_mode(target_hvac_mode)
+
+        if target_temperature is not None:
+            # single setpoint control
```
```suggestion
```

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -408,6 +409,89 @@ def _get_temperature_in_degrees(
+    def _update_from_device(self) -> None:
+        """Update from device."""
+        self._calculate_features()
+        self._attr_hvac_action = None
```
```suggestion
```

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -408,6 +409,89 @@ def _get_temperature_in_degrees(
+        if self._feature_map == feature_map:
+            return
+        self._feature_map = feature_map
+        self._attr_hvac_modes: list[HVACMode] = [HVACMode.HEAT_COOL]
```
Add a comment above this line that you can not set the mode for a TemperatureControl cluster, you can only set the target temperature.

Also I'm wondering if heat_cool is the right mode here. Can we somewhere determine if this device is either a cooling or heating device ? I think we can derive that from the device type maybe ? 
So if we know from the devicetype that its a fridge, we should hardcode this mode to cooling

lboue replied:
I set it to HVACMode.HEAT_COOL because TemperatureControlledCabinet can have  cooling/heating functionality:
https://github.com/project-chip/connectedhomeip/blob/master/data_model/1.4/device_types/TemperatureControlledCabinet.xml

Currently `device_type` is only accessible in MatterDiscoverySchema.

lboue replied:
The difficulty here is that it concerns a different EP. In the case of a refrigerator there is 3 EP:
1. Endpoint 1: Device Type(s): Refrigerator
2. Endpoint 2: Device Type(s): Temperature Controlled Cabinet
2. Endpoint 3: Device Type(s): Temperature Controlled Cabinet

![image](https://github.com/user-attachments/assets/9cd91270-44cb-4a7a-ba2e-99ab4eb7cc7d)

marcelveldt replied:
yes you should look at root endpoint if its a refrigerator

lboue replied:
EP0 device type is 'Root Node'
```json
  "attributes": {
    "0/29/0": [
      {
        "0": 22,
        "1": 1
      }
    ],
```

marcelveldt replied:
Sorry for the late reply, I missed this one.
Hmm. so each endpoint has a devicetype of TemperatureControlledCabinet but we also have an endpoint set to devicetype Refigerator or Freezer

So what we need to do find out if its a freezer or fridge, we need to enumerate the endpoints.
Just do that in any setup/init code we may have or a function where we determine features.

But... regardless of that... for no we only have freezers or (combo)fridges and both can only cool.
So why bother now and just always assume cooling for now. By the name the Matter spec/devices ever change and we get a TemperatureControlledCabinet that can be heated, we will have to revisit the code.

lboue replied:
I have made the requested changes. I hope it's OK for you.

marcelveldt commented on homeassistant/components/matter/climate.py:
```diff
@@ -435,4 +519,19 @@ def _get_temperature_in_degrees(
+        required_attributes=(
+            clusters.TemperatureControl.Attributes.TemperatureSetpoint,
+            clusters.TemperatureControl.Attributes.FeatureMap,
+            clusters.TemperatureControl.Attributes.MinTemperature,
+            clusters.TemperatureControl.Attributes.MaxTemperature,
```
are these mandatory ?

lboue replied:
Yes 
![image](https://github.com/user-attachments/assets/99a26f13-58cc-4ee1-9099-f73cb5f335d7)

lboue committed:
Update homeassistant/components/matter/climate.py

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

lboue committed:
Update homeassistant/components/matter/climate.py

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

lboue committed:
Update homeassistant/components/matter/climate.py

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

lboue committed:
Update homeassistant/components/matter/climate.py

Co-authored-by: Marcel van der Veldt <m.vanderveldt@outlook.com>

lboue committed:
Add comment for hvac_mode

lboue committed:
Add TagList attribute

Attributes of Descriptor Cluster
TagList / AttributeId: 4 (0x004) - Value type: Optional[List[Descriptor.Structs.SemanticTagStruct]]
  kNamespaceRefrigerator = 0x41
    kTagRefrigerator
      [{"0":null,"1":65,"2":0}]


    TagFreezer
      [{"0":null,"1":65,"2":1}]

lboue committed:
Testing TagList to autodetect device type

joostlek reviewed (changes requested):
The CI is failing, can you take a look?

lboue commented:
> The CI is failing, can you take a look?

Yes, I'm waiting for an answer from @marcelveldt  on how to deal with this: https://github.com/home-assistant/core/pull/128838#discussion_r1942718411

lboue committed:
Merge branch 'dev' into TemperatureControlled

lboue committed:
Remove TagList

lboue committed:
Update tests

lboue committed:
Switch to HVACMode.COOL as requested

lboue committed:
Resolving conflicts

lboue committed:
Merge branch 'dev' into TemperatureControlled

lboue committed:
Fix tests

lboue committed:
Update snashots

lboue commented:
> The CI is failing, can you take a look?

I've updated the snapshots. The CI is OK now.

marcelveldt reviewed (approved):
Looks good to me. Nice work @lboue

lboue commented:
> The CI is failing, can you take a look?

Is it OK now?

MartinHjelmare commented:
None of the mentioned devices in the PR description are meant to be represented with a climate entity which has HVAC mode as the main state. A number entity would be ok.
